external_executors/dnanexus_wes/dx-wes-lambda/dx_wes.py

```diff
@@ -0,0 +1,510 @@
+
+POST /workflows                       dnanexus_workflows_post
+GET  /workflows/{workflow_id}         dnanexus_workflows_get
+GET  /workflows/{workflow_id}/status  dnanexus_workflows_get_status
+"""
+
+import datetime
+import functools
+import json
+import os
+import shutil
+import subprocess
+import sys
+import uuid
+
+from multiprocessing.dummy import Pool
+
+import boto3
+import dxpy
+import google.cloud.storage
+import requests
+
+##############################################################################
+## Helper functions
+##############################################################################
+
+def set_dx_authorization(authorization_header):
+    """Set up authorization to DNAnexus. This appears to be done through a
+    module-level global, so we can just do it once.
+
+    Return the auth_header suitable for inserting into the environment of
+    a subprocess.
+    """
+
+    dx_token = authorization_header.replace("Bearer ", "")
+    auth_header = {
+        "auth_token_type": "Bearer",
+        "auth_token": dx_token
+    }
+    dxpy.set_security_context(auth_header)
+    return json.dumps(auth_header)
+
+def _lookup_table(wes_workflow_id, event_type, field):
+    """Lookup a value in the DynamoDB table.
+
+    event_type is the EventType field in the query
+    field is the column in the one first returned value
+    """
+    ddb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
+    table = ddb.Table(os.environ["AWS_DDB_TABLE"])
+    response = table.query(
+        KeyConditionExpression=boto3.dynamodb.conditions.Key('WESWorkflowId').eq(wes_workflow_id) & \
+                               boto3.dynamodb.conditions.Key("EventType").eq(event_type)
+    )
+
+    # If Count is zero, then there's no recorded analysis id so just return
+    # None
+    if not response["Count"]:
+        return None
+
+    value = response["Items"][0][field]
+
+    return value
+
+lookup_dx_analysis_id = functools.partial(_lookup_table, event_type="DxAnalysisId", field="Id")
+lookup_dxwdl_log = functools.partial(_lookup_table, event_type="DxWDLLog", field="Log")
+
+def record_dxwdl_log(wes_workflow_id, log):
+    """Record a log event associated with the execution of dxWDL itself. For
+    the actual workflow execution logs, we can make requests to DNAnexus, but
+    if reports something interesting, we have to record that ourselved in the
+    database.
+    """
+
+    ddb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
+    table = ddb.Table(os.environ["AWS_DDB_TABLE"])
+    table.put_item(
+        Item={
+            "WESWorkflowId": wes_workflow_id,
+            "EventType": "DxWDLLog",
+            "Log": log
+        }
+    )
+
+def record_dx_analysis_id(wes_workflow_id, dx_analysis_id):
+    """Record an association between a WES workflow id and a DXAnalysis id."""
+
+    ddb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
+    table = ddb.Table(os.environ["AWS_DDB_TABLE"])
```

Comment:
If this is the table you're logging everything to, I wonder if `AWS_DDB_TABLE` should be a bit more expressive about the contents of the table. `AWS_DXWDL_TABLE` etc...

Reply:
Yes, will add some comments about that.